### PR TITLE
ceph-detect-init: detect init system by poking the system

### PR DESCRIPTION
**Pull request by tchaikov**
Fixes: http://tracker.ceph.com/issues/19884

### CODE REVIEW

**Comment by tchaikov** [comment]
@dachary would you please take a look?

**Comment by dachary** [comment]
This is tricky because we don't actually have integration tests for those platforms. It may be better to be very conservative and only add support for the missing operating system in the same way it was done before. IIRC a pitfall is that some operating system versions can have two init system installed at the same time. But ceph has only been tested on one of them. That being said, since we don't actually have teuthology tests to verify Ceph deploys and run, it's hard to tell what should work and what does not. Reason why I'd suggest a minimal change instead of a (sensible but uncertain) refactor. Does that make sense ?

**Comment by gaudenz** [comment]
@dachary as Debian supports both sysvinit and systemd in at least jessie (current stable) and stretch (soon to be released new stable) the current logic which hardcodes one init system per Debian version just does not make sense. I much prefer the approach suggested in this PR.
The hardcoding would make at least some sense from Ceph's POV if it refused to run under sysvinit when installed on jessie or stretch, but the current packages still ship sysvinit scripts for these releases. And from a Debian POV this is a good thing as packages are supposed to still support sysvinit.

**Comment by gaudenz** [comment]
@tchaikov Probably you should also remove the release name detection code in ceph_detect_init/__init__.py (lines 143-163) alltogether to not create any confusion as this is no longer used by your code.

**Comment by dachary** [comment]
> Debian supports both sysvinit and systemd in at least jessie (current stable) and stretch (soon to be released new stable)

My point is that Ceph (not Debian) has only been tested with either sysvinit or systemd on Debian, not both. I'm not even sure where to look to find traces of the last integration tests regarding these.

**Comment by tchaikov** [comment]
@dachary i think the problem is two folded with the precondition of: we want to be sure that Ceph behave on all supported distros/releases.

with this change, user could be using Ceph with untested combinations, like wheezy + systemd (systemd was introduced to wheezy as a technical preview) or jessie + sysvinit, and runs into unexpected issues. because we don't test Ceph on Debian or its derivative with these unknown combinations. so our user could run into unexpected problems. but what if he/she is able to take the risk, and really **wants** to use Ceph with the untested setting (jessie + sysvinit, for instance)? shall we error out in that case? IMHO, it would be better if we **allow** them to do it. for example, if rados is able to compile and run on OpenIndiana, shall we stop our user from using librados on that platform because we don't have any teuthology integration test for that combination?

when it comes to `ceph-disk` and `ceph-detect-init`, i think, probably we don't need to forbid the untested combinations. but instead, to put a warning in the README or release document or man page, so user is aware of the risk of experiencing undefined behavior.

in other words, if a user runs `ceph-detect-init` on jessie with only sysvinit installed, `sysvinit` is returned. if a user runs `ceph-detect-init` a system with both `sysvinit` **and** `systemd` installed, but only one of them is activated, we return the activated one, even if it is not tested in our integration test.

what do you think?

**Comment by tchaikov** [comment]
changelog

- remove not used code and tests 

@gaudenz fixed and repushed

**Comment by dachary** [comment]
@tchaikov this is a sane approach, good idea :-)

**Comment by tchaikov** [comment]
@dachary thanks =) i added a "Bugs" section in the related manpages.

**Comment by gaudenz** [comment]
@tchaikov @dachary Just as a note to anyone wanting to backport this to jewel: You also need at least the first part of commit e442f564195120db0595753ba355b9f0f8156521 for the detection to work on Debian stretch. The Python platform detection code returns '' as the codename on Debian stretch.

**Comment by gaudenz** [comment]
And of course it would be nice to have this backported to jewel as many people will still be running jewel at the point where they upgrade to stretch.

**Comment by tchaikov** [comment]
@gaudenz i marked the tracker ticket so the it can be backported after being resolved.

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-05-16_22:22:40-rados-wip-sage-testing2---basic-smithi/
